### PR TITLE
Verify transaction root and ommers hash in header before state transitioning

### PR DESCRIPTION
**Pull request by Bhargavasomu**
- Improve tests so that it verifies rlp and hash values for
	  blocks and headers
	- Verify state post state transition by blocks

### What was wrong?
Transaction Root and Ommers hash weren't being verified before state transitioning. This PR fixes that.
Related to Issue #12 

### How was it fixed?
Fixed based on Trie implementation

#### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: src/eth1spec/rlp.py
Comment (SamWilsn) [inline src/eth1spec/rlp.py]: Interesting! Does this also handle nested dataclasses?

Comment (SamWilsn) [reply src/eth1spec/rlp.py]: In #37, @voith [implemented RLP encoding differently](https://github.com/quilt/eth1.0-specs/pull/37/files#diff-870181b4345e7de97307aa360f3958e009891c4019204a09ce59e8b4f6db0038R351) for block headers.

I don't want to express an opinion either way, just wanted to point out that there are competing styles and we should decide on one.

Comment (SamWilsn) [reply src/eth1spec/rlp.py]: I've been thinking about this a bit, and I'm pretty sure I prefer the explicit style of encoding. For example, something like:

```python
def account_to_tuple(account: Account) -> Tuple[RLP]:
    return (
        account.nonce,
        account.balance,
        root(map_keys(account.storage)),
        crypto.keccak256(account.code),
    )
```

Thoughts?

File: src/eth1spec/rlp.py
Comment (SamWilsn) [inline src/eth1spec/rlp.py]: Similarly, I think decoding should be in an `account_from_tuple(decoded: Tuple[RLP])`, but I could certainly be convinced otherwise.

**Review by SamWilsn — CHANGES_REQUESTED** [review]
Looks pretty good. I think we need to sort out the RLP-encoding of dataclasses before merging though.

File: src/eth1spec/rlp.py
Comment (Bhargavasomu) [reply src/eth1spec/rlp.py]: I actually also want the explicit encoding and decoding. I'll try a bit more to add this explicit encoding/decoding feature.

File: src/eth1spec/rlp.py
Comment (Bhargavasomu) [reply src/eth1spec/rlp.py]: I have currently not implemented the decoding to a dataclass like Account, etc. as we aren't using the decoding anywhere now. Regardless I would have implemented this, except for the fact that on decoding the `Account` rlp encoding, we get the `code_hash` and `storage_root` instead of `code` and `storage` which is what is stored in the dataclass, which is causing an inconsistency wrt the other dataclasses. Should we consider restructuring the `Account` dataclass?

\cc @lightclient

File: src/eth1spec/rlp.py
Comment (lightclient) [inline src/eth1spec/rlp.py:424]: I'm a little worried to call this `encode_account` because it masks the patricialization happening. I think it is okay for now, but we'll definitely want to change in the future.

File: src/eth1spec/rlp.py
Comment (lightclient) [reply src/eth1spec/rlp.py]: I think at some point we may want to create a dataclass like `RawAccount` or something that has all immediate. For now, I think prefer to just remove this decode function.

**Review by lightclient — APPROVED** [review]
This looks good to me. I think the only 2 things that would be nice to do before merging is to:

1. Note that `encode_account` does more than just RLP encoding and add a `TODO` that we'll simplify this in the future.
2. Remove the decoder function stub.

File: src/eth1spec/rlp.py
Comment (Bhargavasomu) [reply src/eth1spec/rlp.py]: Makes sense.

**Comment by Bhargavasomu** [comment]
@lightclient @SamWilsn I believe I have addressed the comments. Can you please merge the PR? Thanks.

**Comment by lightclient** [comment]
Look good to me, @SamWilsn PTAL and merge if it's good on your end.

File: src/eth1spec/rlp.py
Comment (SamWilsn) [inline src/eth1spec/rlp.py:76]: We've got two options here:
 - Either we need to add `Block`, `Header`, etc to the definition of the `RLP` type; or
 - We only handle tuples in `encode`, and create separate functions (ex. `block_to_tuple(...) -> Tuple[RLP]`) for each type.

**Review by SamWilsn — CHANGES_REQUESTED** [review]

File: src/eth1spec/rlp.py
Comment (Bhargavasomu) [reply src/eth1spec/rlp.py:76]: Oops, I forgot to add `Block` and `Header` to the definition of the `RLP` type. Solution 2 sounds good as well. I'm guessing that this would be rearchitected when we implement decode to the dataclasses.

**Review by SamWilsn — APPROVED** [review]

**Pull request merged**
